### ISSUE TITLE
Get a thread safe copy of torch::nn::Sequential object

### ISSUE THREAD

**Issue by oroojlooy** [pytorch/pytorch#54874]
I need to get a copy of `shared` neural network of type `torch::nn::Sequential`. It seems that there is no available API for this purpose at the moment. It seems that declaring and instantiating the new model as `auto l_model = torch::nn::Sequential(model);` should help, although it still does not work.  

This can be useful feature specially for the multi-threaded implementation of algorithms. In each thread, one may need to change the weights of the neural net and that requires to make sure that the weights do not mix up with each other among the threads. It seems that one cannot directly use `auto local_model = global_model;` since it only provides a new pointer which points to the same address in the memory that the original object lives in. So, whenever one changes the weights in one thread, all other threads see that change, which is not desirable. 

Here is an example that shows a use-case:

    #include <omp.h>
    #include "iostream"
    #include <ATen/ATen.h>
    #include <torch/torch.h>
    
    void set_weights(torch::nn::Sequential &model, float w){
    
        torch::autograd::GradMode::set_enabled(false);
        for (auto &p : model->named_parameters()) {
            std::string y = p.key();
            auto z = p.value(); // note that z is a Tensor, same as &p : layers->parameters
    
            if (y.compare(2, 6, "weight") == 0)
                z.uniform_(w, w);
            else if (y.compare(2, 4, "bias") == 0)
                z.uniform_(0, 0);
        }
        torch::autograd::GradMode::set_enabled(true);
    
    }
    
    void print_weights(torch::nn::Sequential &model) {
        //Returns an iterator over module parameters, yielding both the name of the parameter as well as the parameter itself.
        auto params = model->named_parameters(true /*recurse*/);
        for (const auto &pair : params) {
            std::cout << pair.key() << ": " << pair.value() << std::endl;
        }
    }
    
    int main()
    {
        int N=3;
        torch::Tensor x = torch::tensor({1,1,1,1}).toType(torch::kFloat32);
        int thread;
    
        torch::nn::Sequential model(torch::nn::Linear(4,2),
                                    torch::nn::Functional(torch::relu),
                                    torch::nn::Linear(2,1));
        torch::Device device_ = torch::Device(torch::kCPU);
        model->to(device_);
        model->to(torch::kFloat32);
        set_weights(model, 1.);
        omp_set_num_threads(2);
    #pragma omp parallel for private(thread) shared(N) //, private(x)
        for (int i=0; i<N; i++) {
            thread = omp_get_thread_num();
            auto l_model = model;
    
            std::cout << "g-model" <<std::endl;
            print_weights(model);
            set_weights(l_model, (float)thread);
            std::cout << "l-model" <<std::endl;
            print_weights(l_model);
            std::cout << "g-model" <<std::endl;
            print_weights(model);
            auto out = l_model->forward(x);
            auto tmp = out.item<float>();
            result[i][0] = (float)thread;
            result[i][1] = *out_f;
        }
        for (int i=0; i <N ; i++){
            std::cout << "i=" << i << " thread=" << result[i][0] << " out=" << result[i][1] << std::endl;
        }
    
        return 0;
    }

**Comment by H-Huang** [comment]
Perhaps I am misunderstanding something, but can't this be accomplished with standard C++ synchronization primitives without changing the API? I also am not understanding the benefit because using a thread-safe implementation means that only one thread can train the model at a time which defeats the purpose of parallelizing it.

I would recommend looking into DataParallel and DistributedDataParallel (https://pytorch.org/tutorials/beginner/dist_overview.html). The basic idea is to replicate the model across multiple GPUs (or multiple machines) that run independently from each other, except for the gradient synchronization and communication which is handled for you.

**Comment by oroojlooy** [comment]
Let me first answer your second question about why this can be useful. I want this feature to implement parallelized RL algorithm, like A2C/A3C. In this kind of algorithms, two cases may happen: 
1- Each thread uses the `global-model`, runs the rollout which basically interacts with the environment and gathers some observations. Then, all the observations are passed back to the main thread and it takes one gradient step. A2C can be implemented in this way. 
2- Each thread takes a copy of the `global-model`, lets call it `local-model`. It runs the rollout by the using the `local-model`, gathers the observations, obtains the gradient, and pass the gradients to the main thread. The main thread uses the gradient, updates the weights of the `global-model` and returns the new weights to the local thread. A3C can be implemented in this way. 

For algorithms like A2C and A3C, one need a multi-thread implementation, if the goal is to use a single machine. 

About your first question, I am not sure which `standard C++ synchronization primitives` features are you referring to. I searched through the APIs and I could not find something like this. In the python side, `torch.multiprocessing` which is a wrapper over `multiprocessing` package handles this need, but I could not find a similar functionality in the `C++` side. If there is such an API, would you please send a link to the documentation of that API?

**Comment by H-Huang** [comment]
Thanks for clarifying! By standard synchronization primitives I meant locks, mutex, semaphores, etc.

Your use case can be handled using Distributed RPC, which allows data to be referenced remotely which you can write in such a way to spin up multiple actors and orchestrate them so that they update a shared network asynchronously. You can see an example for RL here (https://pytorch.org/tutorials/intermediate/rpc_tutorial.html?highlight=reinforcement%20learning)

I also recommend looking into rlstructures (https://github.com/facebookresearch/rlstructures) which provides a framework specifically for RL and includes RL algorithms built in, such as A2C.

I don't believe there are any plans to make this API thread safe as our current packages handles these use cases. Please take a look at these libraries and try them out. If there are any gaps / concerns feel free to raise them!

**Comment by H-Huang** [comment]
We don't have an official C++ API for distributed RPC, the majority of the implementation is in C++, but our APIs are provided in python.

**Comment by oroojlooy** [comment]
Thanks for the information. 
Back to the first question, how can I get a deep copy of `torch::nn::Sequential()` object? Basically, I need to declare an instance of `torch::nn::Sequential()` and copy an existing network to that. For sure, I can write a new copy function for copying each element one by one, but If I could access the underlying object, I could just copy that object and assign its address to a new pointer.
